src/driver/rtc.c

```diff
@@ -0,0 +1,266 @@
+static const char *RTC_DEV = "/dev/rtc";
+static const char *RTC_FILE = "/mnt/extsd/rtc.txt";
+static const unsigned char RTC_DAYS_PER_MONTH[] = {
+    31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31};
+
+/**
+ *  Globals
+ */
+static int g_rtc_has_battery = 0;
+
+/**
+ *  Returns 1 if leap year is detected, otherwise 0.
+ */
+static inline int rtc_is_leap_year(unsigned int year) {
+    return (!(year % 4) && (year % 100)) || !(year % 400);
+}
+
+/*
+ * The number of days per month.
+ */
+static inline int rtc_days_per_month(unsigned int year, unsigned int month) {
+    return (rtc_is_leap_year(year) && month == 1) + RTC_DAYS_PER_MONTH[month];
+}
+
+/*
+ * Returns 0 if rtc_date is valid, otherwise -EINVAL (-22).
+ */
+int rtc_has_valid_date(struct rtc_date *rd) {
+    return (rd->year - 1900 < 70 ||
+            rd->month - 1 >= 12 ||
+            rd->day > rtc_days_per_month(rd->year, rd->month - 1) ||
+            rd->hour >= 24 ||
+            rd->min >= 60 ||
+            rd->sec >= 60)
+               ? -EINVAL
+               : 0;
+}
+
+/**
+ *  Return 0 when battery detected, otherwise -ENODEV (-19).
+ */
+int rtc_has_battery() {
+    return g_rtc_has_battery ? 0 : -ENODEV;
+}
+
+/**
+ *  Log RTC date structure in a standard format
+ */
+void rtc_print(const struct rtc_date *rd) {
+    LOGI(RTC_LOG_FORMAT,
+         rd->year,
+         rd->month,
+         rd->day,
+         rd->hour,
+         rd->min,
+         rd->sec);
+}
+
+/**
+ *  Conversion Utils used to interact with internal data types.
+ */
+void rtc_rd2rt(const struct rtc_date *rd, struct rtc_time *rt) {
+    rt->tm_year = rd->year - 1900;
+    rt->tm_mon = rd->month - 1;
+    rt->tm_mday = rd->day;
+    rt->tm_hour = rd->hour;
+    rt->tm_min = rd->min;
+    rt->tm_sec = rd->sec;
+}
+void rtc_rt2rd(const struct rtc_time *rt, struct rtc_date *rd) {
+    rd->year = rt->tm_year + 1900;
+    rd->month = rt->tm_mon + 1;
+    rd->day = rt->tm_mday;
+    rd->hour = rt->tm_hour;
+    rd->min = rt->tm_min;
+    rd->sec = rt->tm_sec;
+}
+
+/**
+ *  Conversion Utils used to interact with system data types.
+ */
+void rtc_rd2tv(const struct rtc_date *rd, struct timeval *tv) {
+    struct rtc_time rt;
+    rtc_rd2rt(rd, &rt);
+    tv->tv_sec = mktime((struct tm *)&rt);
+    tv->tv_usec = 0;
+}
+void rtc_tv2rd(const struct timeval *tv, struct rtc_date *rd) {
+    struct rtc_time *rt = (struct rtc_time *)gmtime(&tv->tv_sec);
+    rtc_rt2rd(rt, rd);
+}
+
+/**
+ *  Initialize RTC and OS Clock from file if detected,
+ *  contents of file must match the following:
+ *  YYYY-MM-DDTHH_MM_SS
+ *  2023-03-10T01_05_15
+ */
+void rtc_init() {
+    struct rtc_date rd_now;
+    rtc_get_clock(&rd_now);
+
+    // Has time has accumulated since the
+    // the installation of the battery?
+    g_rtc_has_battery =
+        !(rd_now.year == 1970 &&
+          rd_now.month == 1 &&
+          rd_now.day == 1 &&
+          rd_now.hour == 0 &&
+          rd_now.min == 0);
```

Comment:
this check currently produces false positives if the application is launched a > min after the goggles have booted.
shouldn't be a problem in regular use, but occurs quite frequently during development.
i propose we re-write to simply `g_rtc_has_battery = rd_now.year > 1970;`

Reply:
Yeah there really is no good way to check if the battery was installed.... especially after the initial installation of the battery.  I will make this change to simplify it.